core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java

```diff
@@ -0,0 +1,224 @@
+        }
+    }
+
+    /**
+     * Parses the active shard count from the given string.  Valid values are "all" for
+     * all shard copies, null for the default value set by {@link IndexSettings#WAIT_FOR_ACTIVE_SHARDS_SETTING}
+     * (which defaults to one shard copy), or a numeric value greater than or equal to 0.
+     * Any other input will throw an IllegalArgumentException.
+     */
+    public static ActiveShardCount parseString(final String str) {
+        if (str == null) {
+            return ActiveShardCount.DEFAULT;
+        } else if (str.equals("all")) {
+            return ActiveShardCount.ALL;
+        } else {
+            int val;
+            try {
+                val = Integer.parseInt(str);
+            } catch (NumberFormatException e) {
+                throw new IllegalArgumentException("cannot parse ActiveShardCount[" + str + "]");
+            }
+            return ActiveShardCount.from(val);
+        }
+    }
+
+    /**
+     * Returns true iff the given cluster state's routing table contains enough active
+     * shards to meet the required shard count represented by this instance.
+     */
+    public boolean enoughShardsActive(final ClusterState clusterState, final String indexName) {
+        if (this == ActiveShardCount.NONE) {
+            // not waiting for any active shards
+            return true;
+        }
+        final IndexMetaData indexMetaData = clusterState.metaData().index(indexName);
+        if (indexMetaData == null) {
+            // its possible the index was deleted while waiting for active shard copies,
+            // in this case, we'll just consider it that we have enough active shard copies
+            // and we can stop waiting
+            return true;
+        }
+        final ActiveShardCount waitForActiveShards;
+        if (this == ActiveShardCount.DEFAULT) {
+            waitForActiveShards = WAIT_FOR_ACTIVE_SHARDS_SETTING.get(indexMetaData.getSettings());
+        } else {
+            waitForActiveShards = this;
+        }
+        final IndexRoutingTable indexRoutingTable = clusterState.routingTable().index(indexName);
+        assert indexRoutingTable != null;
+        if (indexRoutingTable.allPrimaryShardsActive() == false) {
+            // all primary shards aren't active yet
+            return false;
+        }
+        for (final IntObjectCursor<IndexShardRoutingTable> shardRouting : indexRoutingTable.getShards()) {
+            if (waitForActiveShards.checkWaitForActiveShards(shardRouting.value, indexMetaData) == false) {
+                // not enough active shard copies yet
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Returns true iff the active shard count in the shard routing table is enough
+     * to meet the required shard count represented by this instance.
+     */
+    public boolean checkWaitForActiveShards(final IndexShardRoutingTable shardRoutingTable, final IndexMetaData indexMetaData) {
```

Comment:
I would give this method the same name as the method above: `enoughShardsActive`. 
One operates at index, the other at the shard level. To make things even more explicit, the above method could have parameters
(IndexRoutingTable, IndexMetaData) and this one (IndexShardRoutingTable, IndexMetaData). This would also leave it up to the caller to deal with `indexMetaData == null`.